File: experiments/random/randomness/spectral/spectral.py
"""
Experiment: Spectral Test of Randomness.
"""

from core.random.rndgen import MarcianiMultiStream as RandomGenerator
from core.random.randomness import spectral as test
from core.utils.report import SimpleReport
from core.utils.file_utils import save_list_of_pairs
from os import path

# Generator
DEFAULT_GENERATOR = RandomGenerator()

# Test Parameters
DEFAULT_SAMSIZE = 100000

# Results directory
DEFAULT_OUTDIR = "out"


def experiment(generator, samsize, outdir):

    filename = path.join(outdir, "spectral_{}_{}_{}".format(generator.__class__.__name__, generator.get_initial_seed(), samsize))

    # Observations: [(u1, u2)]
    data = test.observations(DEFAULT_GENERATOR.rnd, DEFAULT_SAMSIZE)

    # Save raw data
    save_list_of_pairs(filename + "csv", data)

    # Report
    r = SimpleReport("SPECTRAL TEST")
    r.add("Generator", "Class", generator.__class__.__name__)
    r.add("Generator", "Seed", generator.get_initial_seed())
    r.add("Test Parameters", "Sample Size", samsize)

    r.save(filename + ".txt")

    print(r)

    # Plot
    #fig_filename = "{}.{}".format(filename, PLT_EXT)
    #test.plot(data, filename=fig_filename)

    # Plot (with zoom)
    #fig_filename = "{}-zoom.{}".format(filename, PLT_EXT)
    #test.plot(data, filename=fig_filename, zoom=(0.5, 0.6))


if __name__ == "__main__":
    experiment(DEFAULT_GENERATOR, DEFAULT_SAMSIZE, DEFAULT_OUTDIR)

File: experiments/random/randomness/spectral/spectral.m
% %
% SPECTRAL TEST
% %

% VARIABLES
modulus = 2147483647;
multiplier = 16807;
%multiplier = 48271;
%multiplier = 50812;
samsize = modulus-1;
a = 0.0;
b = 0.001;

% DATASET
filename = sprintf("out/mod%d_mul%d/spectral_%d.csv", modulus, multiplier, samsize);

t = readtable(filename);

x = t{:,1:1};
y = t{:,2:2};

% PLOT
figure(1);
scatter(x, y);
xlim([a b])
ylim([a b])
title({"Spectral Test";sprintf("Modulus: %d | Multiplier: %d", modulus, multiplier)});
xlabel("Random Number");
ylabel("Random Number");
